synapseutils/copy.py

```diff
@@ -2,55 +2,124 @@
-                                                      "newFileName": fileName,
-                                                      "originalFile": {"associateObjectType": associateObjectType,
-                                                                       "fileHandleId": filehandleId,
-                                                                       "associateObjectId": associateObjectId}})
-    copiedFileHandles = syn.restPOST('/filehandles/copy', body=json.dumps(copyFileHandleRequest),
-                                     endpoint=syn.fileHandleEndpoint)
-    _copy_cached_file_handles(syn.cache, copiedFileHandles)
-    return copiedFileHandles
+
+    # Check if length of all inputs are equal
+    if not (len(fileHandles) == len(associateObjectTypes) == len(associateObjectIds)
+            and (newContentTypes is None or len(newContentTypes) == len(associateObjectIds))
+            and (newFileNames is None or len(newFileNames) == len(associateObjectIds))):
+        raise ValueError("Length of all input arguments must be the same")
+
+    # If no optional params passed, assign to empty list
+    if newContentTypes is None:
+        newContentTypes = []
+    if newFileNames is None:
+        newFileNames = []
+
+    # Remove this line if we change API to only take fileHandleIds and not Objects
+    file_handle_ids = [synapseclient.core.utils.id_of(handle) for handle in fileHandles]
+
+    # add division logic for POST call here
+    master_copy_results_list = []  # list which holds all results from POST call
+    num_batches = math.ceil(len(fileHandles) / MAX_FILE_HANDLE_PER_COPY_REQUEST)
+    for i in range(num_batches):
+        start = i * MAX_FILE_HANDLE_PER_COPY_REQUEST
+        end = start + MAX_FILE_HANDLE_PER_COPY_REQUEST
+        batch_copy_results = _copy_file_handles_batch(syn, file_handle_ids[start:end], associateObjectTypes[start:end],
+                                                      associateObjectIds[start:end], newContentTypes[start:end],
+                                                      newFileNames[start:end])
+        list_copy_results = batch_copy_results.get("copyResults")
+        master_copy_results_list.extend(list_copy_results)
+    return master_copy_results_list
+
+
+def _copy_file_handles_batch(self, file_handle_ids, obj_types, obj_ids, new_con_types, new_file_names):
+    #TODO: '''include comment, and validate batch size'''
+    copy_file_handle_request = _create_batch_file_handle_copy_request(file_handle_ids, obj_types, obj_ids,
+                                                                      new_con_types, new_file_names)
+    # make backend call which performs the copy specified by copy_file_handle_request
+    copied_file_handles = self.restPOST('/filehandles/copy', body=json.dumps(copy_file_handle_request),
+                                       endpoint=self.fileHandleEndpoint)
+    return copied_file_handles
```

Comment:
this function can extract the list from the dictionary and return the list instead of the dictionary